File: windows/src/LanStash.Infrastructure/DsmConnectionResolver.cs
using LanStash.Domain;

namespace LanStash.Infrastructure;

public sealed record DiscoveredConnection(
    NasProfile Profile,
    IReadOnlyDictionary<string, ApiCapability> Capabilities,
    DsmConnectionSource Source = DsmConnectionSource.DirectAddress);

/// <summary>
/// 登录前只使用不含凭据的能力发现探测连接候选。
/// 找到可信连接后，调用方才可以提交账号、密码和验证码。
/// </summary>
public sealed class DsmConnectionResolver(
    IDsmApiClient api,
    DsmQuickConnectResolver quickConnect)
{
    public async Task<DiscoveredConnection> DiscoverAsync(
        NasProfile profile,
        Action<string>? updateStatus = null,
        CancellationToken cancellationToken = default)
    {
        var parsed = NasAddressParser.Parse(profile.Host, profile.Port);
        if (parsed.Kind == NasAddressKind.Direct)
        {
            updateStatus?.Invoke(UserText.Key("WinShareda6d7b48863714f5e"));
            var directProfile = profile with { Host = parsed.Host, Port = parsed.Port };
            return new(
                directProfile,
                await api.DiscoverAsync(
                    directProfile,
                    DsmConnectionSource.DirectAddress,
                    cancellationToken),
                DsmConnectionSource.DirectAddress);
        }

        updateStatus?.Invoke(UserText.Key("WinSharedaa0582cad267718e"));
        IReadOnlyList<QuickConnectEndpoint> endpoints;
        try
        {
            endpoints = await quickConnect.ResolveAsync(parsed.Host, cancellationToken);
        }
        catch (DsmException error) when (
            error.Kind == DsmErrorKind.QuickConnectDirectUnavailable)
        {
            endpoints = [];
        }

        DsmException? lastDirectError = null;
        foreach (var endpoint in endpoints)
        {
            updateStatus?.Invoke(endpoint.Kind == QuickConnectEndpointKind.Local
                ? UserText.Key("WinShared3b38866d76d21239")
                : UserText.Key("WinShared307e0c332a164ea1"));
            var connectionProfile = profile with
            {
                Host = endpoint.Host,
                Port = profile.Port ?? endpoint.Port,
            };
            try
            {
                return new(
                    connectionProfile,
                    await api.DiscoverAsync(
                        connectionProfile,
                        ConnectionSourceFor(endpoint.Kind),
                        cancellationToken),
                    ConnectionSourceFor(endpoint.Kind));
            }
            catch (DsmException error)
            {
                lastDirectError = error;
            }
        }

        updateStatus?.Invoke(UserText.Key("WinShared6edbda7d2a81743a"));
        try
        {
            var relay = await quickConnect.RequestRelayAsync(parsed.Host, cancellationToken);
            var relayProfile = profile with { Host = relay.Host, Port = relay.Port };
            return new(
                relayProfile,
                await api.DiscoverAsync(
                    relayProfile,
                    DsmConnectionSource.QuickConnectRelay,
                    cancellationToken),
                DsmConnectionSource.QuickConnectRelay);
        }
        catch (DsmException error) when (
            error.Kind == DsmErrorKind.QuickConnectRelayUnavailable &&
            lastDirectError is not null)
        {
            throw lastDirectError;
        }
    }

    internal static DsmConnectionSource ConnectionSourceFor(
        QuickConnectEndpointKind endpointKind) => endpointKind switch
        {
            QuickConnectEndpointKind.Local => DsmConnectionSource.QuickConnectLan,
            QuickConnectEndpointKind.External => DsmConnectionSource.QuickConnectExternal,
            QuickConnectEndpointKind.Relay => DsmConnectionSource.QuickConnectRelay,
            _ => throw new ArgumentOutOfRangeException(nameof(endpointKind)),
        };
}

File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/network/DsmConnectionResolver.kt
package io.github.qwertyuiop1995.dsmnativeclient.network

import io.github.qwertyuiop1995.dsmnativeclient.domain.ApiCapability
import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmFailure
import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmErrorKind
import io.github.qwertyuiop1995.dsmnativeclient.domain.NasProfile
import kotlinx.coroutines.CancellationException

internal data class DiscoveredConnection(
    val profile: NasProfile,
    val capabilities: Map<String, ApiCapability>,
)

enum class ConnectionStatus {
    PREPARING,
    CONNECTING_DIRECT,
    LOOKING_UP_QUICK_CONNECT,
    TRYING_LOCAL,
    TRYING_EXTERNAL,
    ESTABLISHING_RELAY,
    RESTORING_SESSION,
    AUTHENTICATING,
}

/**
 * 登录前只使用不含凭据的能力发现探测连接候选。
 * 找到可信连接后，调用方才可以提交账号、密码和验证码。
 */
internal class DsmConnectionResolver(
    private val api: DsmApiClient,
    private val quickConnect: DsmQuickConnectResolver = DsmQuickConnectResolver(),
) {
    suspend fun discover(
        profile: NasProfile,
        onStatus: (ConnectionStatus) -> Unit = {},
    ): DiscoveredConnection {
        val parsed = NasAddressParser.parse(profile.address, profile.port)
        if (parsed.kind == NasAddressKind.DIRECT) {
            onStatus(ConnectionStatus.CONNECTING_DIRECT)
            val connectionProfile = profile.copy(
                address = "https://${parsed.host}",
                port = parsed.port,
            )
            return DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
        }

        onStatus(ConnectionStatus.LOOKING_UP_QUICK_CONNECT)
        val endpoints = try {
            quickConnect.resolve(parsed.host)
        } catch (error: DsmFailure) {
            if (error.kind == DsmErrorKind.QUICK_CONNECT_DIRECT_UNAVAILABLE) {
                emptyList()
            } else {
                throw error
            }
        }
        var lastDirectFailure: DsmFailure? = null
        for (endpoint in endpoints) {
            onStatus(
                if (endpoint.kind == QuickConnectEndpointKind.LOCAL) {
                    ConnectionStatus.TRYING_LOCAL
                } else {
                    ConnectionStatus.TRYING_EXTERNAL
                }
            )
            val connectionProfile = profile.copy(
                address = "https://${endpoint.host}",
                port = profile.port ?: endpoint.port,
            )
            try {
                return DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                lastDirectFailure = error
            }
        }

        onStatus(ConnectionStatus.ESTABLISHING_RELAY)
        return try {
            val relay = quickConnect.requestRelay(parsed.host)
            val connectionProfile = profile.copy(
                address = "https://${relay.host}",
                port = relay.port,
            )
            DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
        } catch (error: CancellationException) {
            throw error
        } catch (error: DsmFailure) {
            throw if (error.kind == DsmErrorKind.QUICK_CONNECT_RELAY_UNAVAILABLE) {
                lastDirectFailure ?: error
            } else {
                error
            }
        }
    }
}
